File: frontend/src/services/mprAnalytics.ts
import { api } from './api';

export interface KPIResumo {
    tempo_medio_concepcao_min: number | null;
    tempo_medio_ciclo_completo_min: number | null;
    tempo_medio_parada_of_min: number | null;
    taxa_entrega_no_prazo_pct: number | null;
    taxa_aprovacao_primeira_entrega_pct: number | null;
    taxa_retrabalho_pct: number | null;
    total_ids_solicitadas: number;
    total_ids_entregues: number;
    ofs_impactadas: number;
}

export interface FilaAtivaItem {
    id: string;
    mo_number: string;
    status: string;
    prioridade: string;
    solicitado_em: string | null;
    aging_horas: number;
    responsavel_atual: string | null;
}

export interface VolumePorPeriodoItem {
    label: string;
    solicitadas: number;
    entregues: number;
    no_prazo: int;
}

export interface EvolucaoTempoCicloItem {
    label: string;
    tempo_medio_ciclo_min: number | null;
    tempo_medio_concepcao_min: number | null;
}

export interface RankingResponsaveisItem {
    responsavel_id: number;
    nome: string;
    ids_concluidas: number;
    tempo_medio_concepcao_min: number | null;
    taxa_aprovacao_primeira_pct: number | null;
    ids_em_andamento: number;
}

export interface MotivoRevisaoItem {
    motivo: string;
    quantidade: number;
    percentual: number;
}

export interface ImpactoFabricacaoItem {
    label: string;
    horas_paradas_total: number;
    ofs_afetadas: number;
    tempo_medio_parada_min: number | null;
}

export interface MetadadosResponse {
    motivos_revisao: string[];
    status_options: string[];
    responsaveis: string[];
}

export interface MPRConfigResponse {
    sla_atencao_horas: number;
    sla_critico_horas: number;
}

const buildQuery = (start: string, end: string) => `?periodo_inicio=${encodeURIComponent(start)}&periodo_fim=${encodeURIComponent(end)}`;

export const mprAnalyticsApi = {
    getKPIsResumo: (start: string, end: string): Promise<KPIResumo> =>
        api.get(`/mpr/analytics/kpis/resumo${buildQuery(start, end)}`),

    getFilaAtiva: (): Promise<FilaAtivaItem[]> =>
        api.get('/mpr/analytics/fila-ativa'),

    getVolumePorPeriodo: (start: string, end: string): Promise<VolumePorPeriodoItem[]> =>
        api.get(`/mpr/analytics/volume-por-periodo${buildQuery(start, end)}`),

    getEvolucaoTempoCiclo: (start: string, end: string): Promise<EvolucaoTempoCicloItem[]> =>
        api.get(`/mpr/analytics/evolucao-tempo-ciclo${buildQuery(start, end)}`),

    getRankingResponsaveis: (start: string, end: string): Promise<RankingResponsaveisItem[]> =>
        api.get(`/mpr/analytics/ranking-responsaveis${buildQuery(start, end)}`),

    getMotivosRevisao: (start: string, end: string): Promise<MotivoRevisaoItem[]> =>
        api.get(`/mpr/analytics/motivos-revisao${buildQuery(start, end)}`),

    getImpactoFabricacao: (start: string, end: string): Promise<ImpactoFabricacaoItem[]> =>
        api.get(`/mpr/analytics/impacto-fabricacao${buildQuery(start, end)}`),

    getMetadados: (): Promise<MetadadosResponse> =>
        api.get('/mpr/analytics/metadados'),

    getConfig: (): Promise<MPRConfigResponse> =>
        api.get('/mpr/analytics/config'),

    updateConfig: (updates: { sla_atencao_horas?: number; sla_critico_horas?: number }): Promise<MPRConfigResponse> =>
        api.patch('/mpr/analytics/config', updates),
};

File: backend/app/schemas/mpr_analytics.py
from pydantic import BaseModel, ConfigDict
from typing import Optional, List

class KPIResumoResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    tempo_medio_concepcao_min: Optional[float] = None
    tempo_medio_ciclo_completo_min: Optional[float] = None
    tempo_medio_parada_of_min: Optional[float] = None
    taxa_entrega_no_prazo_pct: Optional[float] = None
    taxa_aprovacao_primeira_entrega_pct: Optional[float] = None
    taxa_retrabalho_pct: Optional[float] = None
    total_ids_solicitadas: int
    total_ids_entregues: int
    ofs_impactadas: int

class FilaAtivaResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    mo_number: str
    status: str
    prioridade: str
    solicitado_em: Optional[str] = None
    aging_horas: float
    responsavel_atual: Optional[str] = None

class VolumePorPeriodoItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    label: str
    solicitadas: int
    entregues: int
    no_prazo: int

class EvolucaoTempoCicloItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    label: str
    tempo_medio_ciclo_min: Optional[float] = None
    tempo_medio_concepcao_min: Optional[float] = None

class RankingResponsaveisItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    responsavel_id: int
    nome: str
    ids_concluidas: int
    tempo_medio_concepcao_min: Optional[float] = None
    taxa_aprovacao_primeira_pct: Optional[float] = None
    ids_em_andamento: int

class MotivoRevisaoItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    motivo: str
    quantidade: int
    percentual: float

class ImpactoFabricacaoItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    label: str
    horas_paradas_total: float
    ofs_afetadas: int
    tempo_medio_parada_min: Optional[float] = None

class MetadadosResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    motivos_revisao: List[str]
    status_options: List[str]
    responsaveis: List[str] # Or objects with id/name

class MPRConfigResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    sla_atencao_horas: int
    sla_critico_horas: int

class MPRConfigUpdate(BaseModel):
    model_config = ConfigDict(from_attributes=True, extra="forbid")
    sla_atencao_horas: Optional[int] = None
    sla_critico_horas: Optional[int] = None
